okhttp/src/main/java/okhttp3/Dispatcher.java

```diff
@@ -75,12 +77,18 @@ public synchronized ExecutorService executorService() {
    * <p>If more than {@code maxRequests} requests are in flight when this is invoked, those requests
    * will remain in flight.
    */
-  public synchronized void setMaxRequests(int maxRequests) {
+  public void setMaxRequests(int maxRequests) {
     if (maxRequests < 1) {
       throw new IllegalArgumentException("max < 1: " + maxRequests);
     }
-    this.maxRequests = maxRequests;
-    promoteCalls();
+    List<AsyncCall> promotedCalls;
+    synchronized (this) {
+      this.maxRequests = maxRequests;
+      promotedCalls = promoteCalls();
+    }
+    for (int i = 0, size = promotedCalls.size(); i < size; i++) {
```

Comment:
I wonder whether this can be simplfied by a promoteAndExecute() call used in multiple places?